File: src/recursion/quick_sort.java
package recursion;

import java.util.Arrays;

public class quick_sort {
    // since we are not passing sub array in the parameters
    // * so we are passing the low and high index to keep track of subarray in
    // recursion
    static void fn(int arr[], int low, int high) {
        // base case
        if (low >= high) {
            return;
        }

        // now we need to swap
        int s = low;
        int e = high;
        // taking middle element as pivot
        int pivot = arr[(s + (e - s) / 2)];

        // * element<pivot will go LHS and elements>pivot will go RHS
        while (s <= e) {
            // if arr is already sorted it will not swap
            while (arr[s] < pivot) {
                s++;
            }
            while (arr[e] > pivot) {
                e--;
            }
            // the above while loop will stop the index at violation
            // * and we will swapt at that violation point
            if (s <= e) {
                // if condition to cope with last step
                int temp = arr[e];
                arr[e] = arr[s];
                arr[s] = temp;
                e--;
                s++;
            }
        }
        // after this swapping end and start will be adjacent to Pivot
        // e P s

        // now pivot as at correct index
        // sort the left half and right half with recursion
        fn(arr, low, e);
        fn(arr, s, high);

    }

    public static void main(String[] args) {
        int arr[] = { 9, 345, 6, 4, 3, 6, 78, 8 };
        fn(arr, 0, arr.length - 1);
        System.out.println(Arrays.toString(arr));
        // Arrays.sort(args);
    }
}

File: typescript/src/Sorting/quickSort.ts
export {};
function quicksort(arr: number[], start: number, end: number) {
  // base case
  if (start >= end) return;

  //   will sort array and give pivot
  // (nums<pivot on LHS of pivot)
  // (nums>pivot on RHS of pivot)
  const pivot: number = partition(arr, start, end);

  // call left half and right half
  quicksort(arr, start, pivot - 1);
  quicksort(arr, pivot, end);
}

function partition(arr: number[], start: number, end: number) {
  // get pivot element (last)
  const pivot = arr[end];
  // i one less than start
  // for left half
  let i = start - 1;

  for (let j = start; j < end; j++) {
    // number < pivot should be on its right side
    if (arr[j] < pivot) {
      i++;
      [arr[i], arr[j]] = [arr[j], arr[i]];
    }
  }
  //   till now pivot is last element
  // final position of pivot (one after i)
  i++;
  [arr[i], arr[end]] = [arr[end], arr[i]];

  return i;
}

const nums = [9, 8, 7, 6, 5, 4, 3, 2, 1];
quicksort(nums, 0, nums.length - 1);
console.log(nums);
